[PasswordManager] Factor out common credentials code.

damienruscoe opened the pull request:
## Description
Adding username and password to a URL is used in multiple places. Factor out this functionality into its own function.

## Motivation and context

## How has this been tested?
Manually

## What is the effect on users?

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [X] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

damienruscoe committed:
[PasswordManager] Factor out common credentials code.

Adding username and password to a URL is used in multiple places. Factor
out this functionality into its own function.

- Use parameter by-value passing which will accept r-values as moves and l-values by copies.
- Explicitly std::move the remaining l-value CURL parameters the call site when the value is not referenced after.
- Rearranged a couple of lines to make the above possible in some cases, to be consistent in others.

howie-f commented on xbmc/utils/URIUtils.h:
```diff
@@ -332,6 +332,9 @@ class URIUtils
    */
   static bool UpdateUrlEncoding(std::string &strFilename);
 
+  static CURL AddCredentials(const CURL& url);
+  static CURL AddCredentials(CURL&& url);
```
out of curiousity. why not only declare the `const CURL&` function and define it like this?
```cpp
CURL URIUtils::AddCredentials(const CURL& url)
{
  CURL authUrl{url};
  if (CPasswordManager::GetInstance().IsURLSupported(authUrl) && authUrl.GetUserName().empty())
    CPasswordManager::GetInstance().AuthenticateURL(authUrl);

  return authUrl;
}
```
should be the same when we never call `URIUtils::AddCredentials(std::move(something));` anywhere or directly with a temporary

howie-f replied:
one copy avoided, got it, thank you

neo1973 commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -1721,3 +1722,15 @@ bool URIUtils::UpdateUrlEncoding(std::string &strFilename)
   return true;
 }
+
+CURL URIUtils::AddCredentials(const CURL& url)
+{
+  return AddCredentials(CURL(url));
+}
+
+CURL URIUtils::AddCredentials(CURL&& authUrl)
+{
+  if (CPasswordManager::GetInstance().IsURLSupported(authUrl) && authUrl.GetUserName().empty())
+    CPasswordManager::GetInstance().AuthenticateURL(authUrl);
+  return authUrl;
+}
```
One single function can handle both cases:
```cpp
CURL URIUtils::AddCredentials(CURL authUrl)
{
  if (CPasswordManager::GetInstance().IsURLSupported(authUrl) && authUrl.GetUserName().empty())
    CPasswordManager::GetInstance().AuthenticateURL(authUrl);
  return authUrl;
}
```

damienruscoe replied:
Genius! The exact semantics I was trying (and failed) to create myself in one function.
Sometimes things are not obvious until it is pointed out :) Thank you!

damienruscoe replied:
@howie-f  Sorry. You are 100% correct. My eyes are getting old! I checked the header file to investigate your concerns and saw the 2 constructors but completely missed the destructor one line below! Disapointment in ones self is the worst kind :/

The destructor is defaulted in the cpp file although it is declared virtual in the header. The virtual adds a new layer of complication! It doesnt make sense to move-from a polymorphic type.

The virtual descructor dates back to 2009 (the migration to git). Ive done a grep of the codebase and not found anything to be deriving from CURL. 

The rule of zero would fit nicely here. Basically remove the virtual destructor and allow it to be implicitly defaulted. Then CURL is a value type with value semantics.

EDIT: The virtual is not the problem as I first though, the declaration of a default destructor (somehow) prevents move semantics on that object. I knocked up a [Godbolt example](https://godbolt.org/z/T11nWr4ee).

damienruscoe replied:
` (include = default or = delete declared)`

This was stated in your post.... My eyes :/

howie-f replied:
> objects have previously had a vtable? Or has the compiler been smart enough to remove it?

no idea, sorry

neo1973 commented on xbmc/filesystem/Directory.cpp:
```diff
@@ -183,17 +183,15 @@ bool CDirectory::GetDirectory(const CURL& url,
-          CPasswordManager::GetInstance().AuthenticateURL(authUrl);
-
-        items.SetURL(url);
+        authUrl = URIUtils::AddCredentials(authUrl);
```
If you want to save on copies then you should use `std::move`:
```suggestion
        authUrl = URIUtils::AddCredentials(std::move(authUrl));
```
There are a couple more locations below  where it would make sense.

damienruscoe replied:
@neo1973 Done.

Do you have any thoughts on trailing moved from types. (An example would be better)

```
void foo() {
  std::vector<int> something;
  use_something( something ); // fine
  move_something( something ); // fine
  // ... 20+ more lines in the function
}
```

I have ensured that `something` is not used after move.... But another developer could come along at a later point and modify one of the 20+ lines (using `something`) without realising the object has been moved from. 

I know it is up to us developers not to make this kind of mistake; but it is also up to us developers to also leave code minimising any future risk.

Interested in hearing your thoughts.

neo1973 reviewed (changes requested):
Good idea to to move that code into a function :+1:

damienruscoe committed:
- Marked CURL class a final enforcing no other class, has been or, can use this as a base class.
- Removed the virtual destructor. This allows the compiler to generate default move/copy assignment/constructors

- Cleaned up comparison operators `CURL::operator==`.
- Allowed missing `CURL == CURL`.
- Also added `string == CURL` to the existing `CURL == string`.
- Added tests for these comparisons.

neo1973 reviewed (approved):
Looks good, thanks :slightly_smiling_face:

howie-f commented on xbmc/URL.cpp:
```diff
@@ -32,7 +32,25 @@
+  return url == Get();
+}
+
+bool operator==(const CURL& url, const std::string& str)
+{
+  return url.Get() == str;
+}
+
+bool operator==(const std::string& str, const CURL& url)
+{
+  return url == str;
+}
```
something looks wrong to me here. both `operator==` are the same (only parameters order is swapped) but the return statements differ.
```
return url.Get() == str;
return url == str;
```

howie-f replied:
maybe better this way? (sorry this is untested and only a suggestion)
```cpp
bool operator==(const CURL& lhs, const std::string& rhs)
{
  return lhs.Get() == rhs;
}

bool operator==(const std::string& lhs, const CURL& rhs)
{
  return operator==(rhs, lhs);
}
```

damienruscoe replied:
No worries. Thanks for the effort you put in :)

howie-f commented on xbmc/URL.cpp:
```diff
@@ -32,7 +32,25 @@
 using namespace ADDON;
 
-CURL::~CURL() = default;
+CURL::CURL(std::string strURL)
+{
+  Parse(std::move(strURL));
+}
+
+bool CURL::operator==(const CURL& url) const
+{
+  return url == Get();
+}
+
+bool operator==(const CURL& url, const std::string& str)
+{
+  return url.Get() == str;
+}
+
+bool operator==(const std::string& str, const CURL& url)
+{
+  return url == str;
+}
```
> Move the implementations back to the header? 
yes, that's what i meant because i'm used to it. but really up to you

damienruscoe replied:
I wish to leave the implementation in the cpp file. Mostly as a "pureist" :) but also a change in implementation doesnt force a huge compilation of dependants of the header file (incremental builds).

damienruscoe commented on xbmc/URL.cpp:
```diff
@@ -32,7 +32,25 @@
+
+bool CURL::operator==(const CURL& url) const
+{
+  return url == Get();
```
Incorrect. Get converts to a string. So the function called in the body will be CURL==std::string which is a different function.

neo1973 commented on xbmc/URL.cpp:
```diff
@@ -32,7 +32,25 @@
+  Parse(std::move(strURL));
+}
+
+bool CURL::operator==(const CURL& url) const
+{
+  return url == Get();
```
Nope, still not true.
